### PR TITLE
mgr/dashboard: Cluster Creation Add Host Section

### PR DESCRIPTION
**Pull request by nizamial09**
**Add host section of the cluster creation workflow.**
Screenshot from 2021-07-05 16-54-44
Screenshot from 2021-07-05 16-55-01

**Routed Modal**
host-add

**TODO** : Implement unit tests for cluster-creation add host section and the wizard component.

Fixes: https://tracker.ceph.com/issues/50565
Fixes: https://tracker.ceph.com/issues/51517

Signed-off-by: Nizamudeen A <nia@redhat.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster.component.html [hunk 1]
```diff
@@ -14,16 +15,49 @@
...
+    <div class="card-body vertical-line">
+      <ng-container [ngSwitch]="currentStep?.stepIndex">
+        <div *ngSwitchCase="'1'"
+             class="ml-5">
+          <h4 class="title"
+              i18n>Add Hosts</h4>
+          <br>
+          <cd-hosts [clusterCreation]="true"></cd-hosts>
+        </div>
+        <div *ngSwitchCase="'2'"
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster.component.html:45]: No need of this here..I'm already introducing it in review section in a later PR.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster.component.ts [hunk 2]
```diff
@@ -1,36 +1,52 @@
...
-  permission: Permission;
-  orchStatus = false;
-  featureAvailable = false;
+export class CreateClusterComponent implements OnDestroy {
+  currentStep: WizardStepModel;
+  currentStepSub: Subscription;
+  permissions: Permissions;
   projectConstants: typeof AppConstants = AppConstants;
+  hosts: Array<object> = [];
+  stepTitles = ['Add Hosts', 'Review'];
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster.component.ts:28]: This line I'm also updating so maybe no need I think..as this PR  just introduces the add-host page `Review` shouldn't be there maybe?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster.component.ts [hunk 3]
```diff
@@ -39,6 +55,60 @@ export class CreateClusterComponent {
...
+  onSubmit() {
+    forkJoin(this.observables)
+      .pipe(
+        finalize(() =>
+          this.clusterService.updateStatus('POST_INSTALLED').subscribe(() => {
+            this.notificationService.show(
+              NotificationType.success,
+              $localize`Cluster creation was successful`
+            );
+            this.router.navigate(['dashboard']);
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster.component.ts:71]: Shouldn't it be  `this.router.navigate(['/dashboard']);` ?

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/wizard-steps.service.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  getSteps(): Observable<WizardStepModel[]> {
+    return this.steps$.asObservable();
+  }
+
+  moveToNextStep(): void {
+    const index = this.currentStep$.value.stepIndex;
+
+    if (index < this.steps$.value.length) {
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/wizard-steps.service.ts:44]: I think if statement not required here and also for `moveToPreviousStep()` , index will always be what is expected in condition so no need to check. I tried and it works without the if statement.

**Review by avanthakkar — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/controllers/host.py [hunk 5]
```diff
@@ -387,7 +387,7 @@ def get(self, hostname: str) -> Dict:
                      'force': (bool, 'Force Enter Maintenance')
                  },
                  responses={200: None, 204: None})
-    def set(self, hostname: str, update_labels: bool = False,
+    def set(self, hostname: str, spec: bool = False, spec_file: Dict = {}, update_labels: bool = False,
```
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/controllers/host.py]: I think these changes are no longer in here.

File: src/pybind/mgr/dashboard/frontend/src/app/app-routing.module.ts [hunk 6]
```diff
@@ -180,6 +181,12 @@ const routes: Routes = [
         component: TelemetryComponent,
         data: { breadcrumbs: 'Telemetry configuration' }
       },
+      {
+        path: 'create-cluster',
```
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/app-routing.module.ts]: These too

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster.component.html (hunk 1 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster.component.html:45]: I know. but I needed it for the NextStep function where I call the host list and update the hosts array using the host list. Apparently calling everything at once in the onSubmit() function is not working as expected.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/wizard-steps.service.ts (hunk 4 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/wizard-steps.service.ts]: Done. Thanks

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster.component.ts (hunk 3 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster.component.ts]: Done.

**Review by epuertat — APPROVED** [review]
LGTM. Thanks @nizamial09 !

**Comment by nizamial09** [comment]
jenkins test dashboard

**Review by avanthakkar — APPROVED** [review]
LGTM!

**Pull request merged**